File: types/model/agent_usage.go
package model

// AgentUsageResponse is the envelope for GET /api/v1/agent/usage. Exactly one
// of Codex / Claude / Antigravity / Kimi / Qoder / Cursor / CodeBuddy is
// populated, matching the requested Type.
type AgentUsageResponse struct {
	Code        int               `json:"code"`
	Type        string            `json:"type"` // "codex" | "claude" | "antigravity" | "kimi" | "qoder" | "cursor" | "codebuddy"
	Codex       *CodexUsage       `json:"codex,omitempty"`
	Claude      *ClaudeUsage      `json:"claude,omitempty"`
	Antigravity *AntigravityUsage `json:"antigravity,omitempty"`
	Kimi        *KimiUsage        `json:"kimi,omitempty"`
	Qoder       *QoderUsage       `json:"qoder,omitempty"`
	Cursor      *CursorUsage      `json:"cursor,omitempty"`
	CodeBuddy   *CodeBuddyUsage   `json:"codebuddy,omitempty"`
}

// UsageWindow is one rate-limit window. LimitWindowSeconds is the source of
// truth for its duration; upstream field names do not imply a fixed period.
type UsageWindow struct {
	UsedPercent        float64 `json:"used_percent"`
	LimitWindowSeconds int64   `json:"limit_window_seconds"`
	ResetAfterSeconds  int64   `json:"reset_after_seconds"`
	ResetAt            int64   `json:"reset_at"` // unix seconds
}

// CodexUsage is the Codex/ChatGPT plan + rate-limit snapshot pulled from the
// ChatGPT usage endpoint.
type CodexUsage struct {
	Email           string       `json:"email,omitempty"`
	PlanType        string       `json:"plan_type,omitempty"`
	Version         string       `json:"version,omitempty"` // effective Codex CLI version used by codex-acp, e.g. "v0.144.0"
	PrimaryWindow   *UsageWindow `json:"primary_window,omitempty"`
	SecondaryWindow *UsageWindow `json:"secondary_window,omitempty"`
	ResetCredits    int          `json:"reset_credits"` // count of available rate-limit reset credits
	// ResetCreditExpiries lists the expiry (unix seconds) of each available reset
	// credit, ascending. Sourced from the rate-limit-reset-credits endpoint;
	// empty when that supplementary call fails.
	ResetCreditExpiries []int64 `json:"reset_credit_expiries,omitempty"`
}

// AgentVersionResponse is the envelope for GET /api/v1/agent/version. It
// reports the installed CLI version of a built-in or custom ACP agent (e.g.
// "v1.17.18"). Version is empty when the binary advertises no parseable version.
type AgentVersionResponse struct {
	Code    int    `json:"code"`
	Version string `json:"version,omitempty"`
}

// ClaudeUsage is the current Claude Code subscription and rate-limit snapshot
// returned by Anthropic's OAuth usage endpoint.
type ClaudeUsage struct {
	PlanType      string                    `json:"plan_type,omitempty"`
	RateLimitTier string                    `json:"rate_limit_tier,omitempty"`
	Version       string                    `json:"version,omitempty"` // effective Claude Code version used by claude-agent-acp, e.g. "v2.1.202"
	FiveHour      *UsageWindow              `json:"five_hour,omitempty"`
	SevenDay      *UsageWindow              `json:"seven_day,omitempty"`
	SevenDayOpus  *UsageWindow              `json:"seven_day_opus,omitempty"`
	WeeklyScoped  []ClaudeScopedUsageWindow `json:"weekly_scoped,omitempty"`
	ExtraUsage    *ClaudeExtraUsage         `json:"extra_usage,omitempty"`
}

// ClaudeScopedUsageWindow is a model-specific weekly limit returned through
// the generic limits array (for example, Fable).
type ClaudeScopedUsageWindow struct {
	Label string `json:"label"`
	UsageWindow
}

// ClaudeExtraUsage describes the optional monthly pay-as-you-go allowance.
type ClaudeExtraUsage struct {
	Enabled      bool     `json:"is_enabled"`
	MonthlyLimit *float64 `json:"monthly_limit,omitempty"`
	UsedCredits  *float64 `json:"used_credits,omitempty"`
	Currency     string   `json:"currency,omitempty"`
}

// AntigravityUsage is the Antigravity (agy) built-in plan snapshot: the agy CLI
// version plus the two model groups' quota windows, each with a 7-day (weekly)
// and a 5-hour bucket. Each window reuses UsageWindow — its UsedPercent is
// derived from the API's remaining fraction, and ResetAt from the bucket's
// reset time. A window is nil when the corresponding bucket is absent.
type AntigravityUsage struct {
	Version      string       `json:"version,omitempty"`       // agy CLI version, e.g. "v1.1.1"
	ClaudeWeekly *UsageWindow `json:"claude_weekly,omitempty"` // Claude/GPT group, 7-day  (bucketId 3p-weekly)
	Claude5h     *UsageWindow `json:"claude_5h,omitempty"`     // Claude/GPT group, 5-hour (bucketId 3p-5h)
	GeminiWeekly *UsageWindow `json:"gemini_weekly,omitempty"` // Gemini group, 7-day  (bucketId gemini-weekly)
	Gemini5h     *UsageWindow `json:"gemini_5h,omitempty"`     // Gemini group, 5-hour (bucketId gemini-5h)
}

// KimiUsage is the Kimi Code plan snapshot from the kimi coding usages
// endpoint: the kimi CLI version plus the three quota windows. Weekly and
// FiveHour are rate-limit windows with a reset time; Total is the cumulative
// quota pool (e.g. purchased credits) and has no reset. A window is nil when
// the API does not report a usable limit for it.
type KimiUsage struct {
	Version       string       `json:"version,omitempty"`        // kimi CLI version, e.g. "v0.1.0"
	ParallelLimit int64        `json:"parallel_limit,omitempty"` // max concurrent sessions
	Weekly        *UsageWindow `json:"weekly,omitempty"`         // 7-day quota (the API's "usage" field)
	FiveHour      *UsageWindow `json:"five_hour,omitempty"`      // 5-hour quota (limits[0].detail)
	Total         *UsageWindow `json:"total,omitempty"`          // cumulative quota (totalQuota), no reset
}

// QoderUsage is the QoderCN credits quota snapshot from the openapi quota
// endpoint. Credits are a single cumulative pool (no rate-limit window reset);
// the pool expires wholesale at ExpiresAt (unix ms from the API, stored here as
// unix seconds). UsedPercent mirrors the API's totalUsagePercentage (0–100).
type QoderUsage struct {
	Version       string  `json:"version,omitempty"`    // qoderclicn CLI version, e.g. "v1.0.48"
	PlanType      string  `json:"plan_type,omitempty"`  // API userType, e.g. "personal_professional_trial"
	Unit          string  `json:"unit,omitempty"`       // quota unit, always "credits"
	Total         float64 `json:"total"`                // total credits in the pool
	Used          float64 `json:"used"`                 // credits consumed
	Remaining     float64 `json:"remaining"`            // credits left
	UsedPercent   float64 `json:"used_percent"`         // 0–100
	ExpiresAt     int64   `json:"expires_at,omitempty"` // unix seconds when the plan/quota expires
	QuotaExceeded bool    `json:"quota_exceeded"`       // true when the pool is exhausted
}

// CursorUsage is the Cursor plan snapshot from cursor.com's usage-summary
// endpoint plus the Cursor Bot (internal codename "Sand") usage RPC, using the
// cursor-agent CLI's locally stored login token. The three plan windows share
// the monthly billing cycle (reset at billingCycleEnd, LimitWindowSeconds =
// cycle length): PrimaryWindow is the plan total, SecondaryWindow the Auto
// lane, TertiaryWindow the API lane. GrokBotWindow is only reported when the
// account has Grok Bot access and the RPC returns a usable percent and reset
// time; otherwise it is nil. A window is nil when the API reports no usable
// percent for it.
type CursorUsage struct {
	Version         string       `json:"version,omitempty"`          // cursor-agent CLI version, e.g. "v2026.09.08"
	MembershipType  string       `json:"membership_type,omitempty"`  // e.g. "pro" | "team" | "enterprise"
	PrimaryWindow   *UsageWindow `json:"primary_window,omitempty"`   // plan total usage (Auto + API lanes)
	SecondaryWindow *UsageWindow `json:"secondary_window,omitempty"` // Auto lane usage
	TertiaryWindow  *UsageWindow `json:"tertiary_window,omitempty"`  // API lane usage
	GrokBotWindow   *UsageWindow `json:"grok_bot_window,omitempty"`  // Grok Bot usage, nil when unavailable
}

// CodeBuddyUsage is the CodeBuddy quota snapshot: the CodeBuddy CLI version,
// plus the monthly quota and the current calendar-month spend in CNY from the
// company Token 看板 OpenAPI (openapi.token.woa.com). QuotaText/CostText keep
// the API's raw string values; Quota is nil when the API reports a non-numeric
// quota ("-" — special states such as unlimited), in which case Remaining and
// UsedPercent are nil too and must not be derived numerically.
type CodeBuddyUsage struct {
	Version     string   `json:"version,omitempty"` // CodeBuddy CLI version, e.g. "v2.6.0"
	Username    string   `json:"username,omitempty"`
	QuotaText   string   `json:"quota_text"`
	CostText    string   `json:"cost_text"`
	Quota       *float64 `json:"quota,omitempty"`
	Cost        *float64 `json:"cost,omitempty"`
	Remaining   *float64 `json:"remaining,omitempty"`
	UsedPercent *float64 `json:"used_percent,omitempty"` // 0–100
}

File: web/src/utils/agentUsage.ts
// Agent subscription / quota info shown on the Home page for supported ACP
// agents. Fetched fresh on every agent-type switch.

export interface UsageWindow {
  used_percent: number;
  limit_window_seconds: number;
  reset_after_seconds: number;
  reset_at: number;
}

export interface CodexUsage {
  email?: string;
  plan_type?: string;
  version?: string; // e.g. "v1.1.0"
  // Upstream field positions; use each window's limit_window_seconds for its duration.
  primary_window?: UsageWindow;
  secondary_window?: UsageWindow;
  reset_credits: number; // count of available rate-limit reset credits
  reset_credit_expiries?: number[]; // unix seconds, one per available credit, ascending
}

export interface ClaudeUsage {
  plan_type?: string;
  rate_limit_tier?: string;
  version?: string; // e.g. "v2.1.202"
  five_hour?: UsageWindow;
  seven_day?: UsageWindow;
  seven_day_opus?: UsageWindow;
  weekly_scoped?: Array<UsageWindow & { label: string }>;
  extra_usage?: {
    is_enabled: boolean;
    monthly_limit?: number;
    used_credits?: number;
    currency?: string;
  };
}

// Antigravity (agy) plan snapshot: the agy CLI version plus the two model
// groups' quota windows (Claude/GPT and Gemini), each with a 7-day (weekly) and
// a 5-hour bucket. Each window reuses UsageWindow — used_percent is the used
// share and reset_at the bucket's reset time. A window is absent when the API
// doesn't report that bucket.
export interface AntigravityUsage {
  version?: string; // e.g. "v1.1.1"
  claude_weekly?: UsageWindow; // Claude/GPT group, 7-day
  claude_5h?: UsageWindow; // Claude/GPT group, 5-hour
  gemini_weekly?: UsageWindow; // Gemini group, 7-day
  gemini_5h?: UsageWindow; // Gemini group, 5-hour
}

// Kimi Code plan snapshot: the kimi CLI version plus the three quota windows.
// weekly / five_hour are rate-limit windows with a reset time; total is the
// cumulative quota pool with no reset. A window is absent when the API doesn't
// report a usable limit for it.
export interface KimiUsage {
  version?: string; // e.g. "v0.1.0"
  parallel_limit?: number; // max concurrent sessions
  weekly?: UsageWindow; // 7-day quota
  five_hour?: UsageWindow; // 5-hour quota
  total?: UsageWindow; // cumulative quota, no reset
}

// QoderCN credits quota snapshot: a single cumulative credits pool (no
// rate-limit window reset). The pool expires wholesale at expires_at.
export interface QoderUsage {
  version?: string; // e.g. "v1.0.48"
  plan_type?: string; // e.g. "personal_professional_trial"
  unit?: string; // always "credits"
  total: number;
  used: number;
  remaining: number;
  used_percent: number; // 0–100
  expires_at?: number; // unix seconds
  quota_exceeded: boolean;
}

// Cursor plan snapshot from cursor.com's usage-summary endpoint plus the Grok
// Bot usage RPC. The three plan windows share the monthly billing-cycle reset
// (reset_at = billingCycleEnd, limit_window_seconds = cycle length):
// primary_window is the plan total, secondary_window the Auto lane,
// tertiary_window the API lane. grok_bot_window is absent when the account has
// no Grok Bot access or the API reports no usable percent/reset for it.
export interface CursorUsage {
  version?: string; // e.g. "v2026.09.08"
  membership_type?: string; // e.g. "pro" | "team" | "enterprise"
  primary_window?: UsageWindow; // plan total usage (Auto + API lanes)
  secondary_window?: UsageWindow; // Auto lane usage
  tertiary_window?: UsageWindow; // API lane usage
  grok_bot_window?: UsageWindow; // Grok Bot usage
}

export type AgentUsageProvider = 'codex' | 'claude' | 'antigravity' | 'kimi' | 'qoder' | 'cursor';

// agentUsageProvider maps a selected built-in agent to a usage provider, or
// null when the agent has no quota view. Match only stable built-in IDs and
// declared historical commands: display names are user-controlled for custom
// agents and must never opt them into another account's quota data.
export function agentUsageProvider(
  agentType?: string,
  _displayName?: string,
): AgentUsageProvider | null {
  const command = (agentType || '').trim().replace(/\s+/g, ' ').toLowerCase();
  if (['antigravity', 'agy', 'antigravity-acp'].includes(command)) return 'antigravity';
  if (['codex', 'codex-acp', 'npx @agentclientprotocol/codex-acp', 'npx @zed-industries/codex-acp'].includes(command)) return 'codex';
  if (['claude', 'claude-agent-acp', 'npx @agentclientprotocol/claude-agent-acp'].includes(command)) return 'claude';
  if (['qoderclicn', 'qoderclicn --acp', 'qwen', 'qwen --acp'].includes(command)) return 'qoder';
  if (['kimi', 'kimi acp'].includes(command)) return 'kimi';
  if (['cursor-agent', 'cursor-agent acp'].includes(command)) return 'cursor';
  return null;
}

export interface AgentUsagePayload {
  codex?: CodexUsage;
  claude?: ClaudeUsage;
  antigravity?: AntigravityUsage;
  kimi?: KimiUsage;
  qoder?: QoderUsage;
  cursor?: CursorUsage;
}

async function readJSONResponse(response: Response, operation: string): Promise<Record<string, unknown>> {
  const body = await response.text();
  let data: Record<string, unknown> | null = null;
  if (body) {
    try {
      const parsed: unknown = JSON.parse(body);
      if (parsed && typeof parsed === 'object' && !Array.isArray(parsed)) {
        data = parsed as Record<string, unknown>;
      }
    } catch {
      // The raw response remains part of the error below.
    }
  }
  if (!response.ok || data?.code !== 0) {
    const detail = body || '(empty response body)';
    throw new Error(`${operation} failed (HTTP ${response.status}): ${detail}`);
  }
  if (!data) {
    throw new Error(`${operation} returned invalid JSON (HTTP ${response.status}): ${body || '(empty response body)'}`);
  }
  return data;
}

export async function fetchAgentUsage(provider: AgentUsageProvider): Promise<AgentUsagePayload> {
  // `cache: 'no-store'` is required: this quota reading changes continuously
  // (Codex windows especially), so a browser/intermediary HTTP-cache hit
  // would serve an old snapshot and — since the result is re-written to the
  // localStorage cache — make the stale value stick across refreshes.
  const url = `/api/v1/agent/usage?type=${provider}`;
  const operation = `GET ${new URL(url, window.location.href).toString()}`;
  let res: Response;
  try {
    res = await fetch(url, { cache: 'no-store' });
  } catch (error) {
    throw new Error(`${operation} failed: ${error instanceof Error ? error.message : String(error)}`);
  }
  const data = await readJSONResponse(res, operation);
  return {
    codex: data.codex as CodexUsage | undefined,
    claude: data.claude as ClaudeUsage | undefined,
    antigravity: data.antigravity as AntigravityUsage | undefined,
    kimi: data.kimi as KimiUsage | undefined,
    qoder: data.qoder as QoderUsage | undefined,
    cursor: data.cursor as CursorUsage | undefined,
  };
}

// fetchAgentVersion returns the installed CLI version of a known ACP agent
// (e.g. "v1.17.18"), keyed by its serve command (the agent's `type`). Used for
// every known agent that has no quota view of its own — the backend resolves
// the command to a binary and runs `<bin> --version`. Returns "" when the agent
// advertises no parseable version; throws on request / unknown-command errors.
export async function fetchAgentVersion(command: string): Promise<string> {
  const url = `/api/v1/agent/version?command=${encodeURIComponent(command)}`;
  const operation = `GET ${new URL(url, window.location.href).toString()}`;
  let res: Response;
  try {
    res = await fetch(url, { cache: 'no-store' });
  } catch (error) {
    throw new Error(`${operation} failed: ${error instanceof Error ? error.message : String(error)}`);
  }
  const data = await readJSONResponse(res, operation);
  return typeof data.version === 'string' ? data.version : '';
}

// Persistent cache of the last successful usage payload per provider, stored in
// localStorage. Lets the card show the previously-fetched plan info instantly —
// on page load, when the user switches agent type, or when the composer
// re-mounts — while a fresh request loads in the background
// (stale-while-revalidate, no loading flash). A failed refresh keeps whatever
// is cached here; with no cache the card shows nothing.
function cacheKey(provider: AgentUsageProvider): string {
  return `agentUsage_${provider}`;
}

export function getCachedUsage(
  provider: AgentUsageProvider,
): CodexUsage | ClaudeUsage | AntigravityUsage | KimiUsage | QoderUsage | CursorUsage | null {
  try {
    const raw = localStorage.getItem(cacheKey(provider));
    if (!raw) return null;
    const obj = JSON.parse(raw);
    return obj && typeof obj === 'object' ? obj : null;
  } catch {
    return null;
  }
}

export function setCachedUsage(provider: AgentUsageProvider, data: AgentUsagePayload): void {
  const value = data[provider];
  try {
    if (value) localStorage.setItem(cacheKey(provider), JSON.stringify(value));
    else localStorage.removeItem(cacheKey(provider));
  } catch {
    /* ignore quota / serialization errors */
  }
}

// Persistent cache of the last successful version string per agent command,
// keyed by the serve command so switching agents shows the previously-probed
// version instantly while a fresh probe runs in the background.
function versionCacheKey(command: string): string {
  return `agentVersion_${command}`;
}

export function getCachedVersion(command: string): string {
  try {
    return localStorage.getItem(versionCacheKey(command)) || '';
  } catch {
    return '';
  }
}

export function setCachedVersion(command: string, version: string): void {
  try {
    if (version) localStorage.setItem(versionCacheKey(command), version);
    else localStorage.removeItem(versionCacheKey(command));
  } catch {
    /* ignore quota / serialization errors */
  }
}
